### PR TITLE
[FSDP2] Added `set_reduce_scatter_divide_factor`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129286

This PR adds an API `FSDPModule.set_reduce_scatter_divide_factor` to allow setting a custom gradient divide factor for reduce-scatter. This can be useful when using parallelisms in combination with FSDP (e.g. expert parallelism), where gradients need to be divided by a custom factor (e.g. an extra `EP` factor).

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 1]
```diff
@@ -329,6 +330,21 @@ def set_post_optim_event(self, event: torch.cuda.Event) -> None:
...
+        custom reduce op using NCCL's PreMulSum, which allows multiplying by
+        the factor before reduction.
+
+        Args:
+            factor (float): Custom divide factor.
+        """
+        state = self._get_fsdp_state()
+        if (fsdp_param_group := state._fsdp_param_group) is not None:
+            mul_factor = 1.0 / float(factor)
+            reduce_op = _make_nccl_premul_sum(mul_factor)
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/fully_shard.py:345]: Nice! A few questions:
- This doesn't yet work with compile right?
- Is there any case where we want to favor `_div_if_needed` over `premul_sum` when nccl is used?

Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:345]: I am not sure about compile (have not tried).

I think we always prefer `premul_sum` for the pre-divide over `_div_if_needed`. I just did not know about this feature until I wrote this PR (which is for some internal use case exploration) 😄 .

Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:352]: I updated the PR to get rid of `_div_if_needed` before reduction for fp16 path.

Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:352]: Nevermind. MTPG does not support premulsum, and the Python plumbing is not there yet to support this. I will remove it from this PR for now.

**Review by weifengpy — APPROVED** [review]

**Comment by awgu** [comment]
[pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 4, 5, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/129286#27817167624) ([gh](https://github.com/pytorch/pytorch/actions/runs/10062525253/job/27817167624))
profiler test failure looks unrelated

**Comment by awgu** [comment]
Screen Shot 2024-07-23 at 9 59 58 PM
rocm tests queueing for 9 hours is tough 😢

**Pull request closed without merging**